_posts/2022-08-10-architecting-trino-next-gen-etl-engine.md

```diff
@@ -0,0 +1,167 @@
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/current-architecture.png)
+
+To address this, we introduce a new execution paradigm to Trino by adding
+exchange spooling that stores intermediate data in an efficient buffering layer
+at stage boundaries.This opens up a lot of new opportunities around failure
+recovery, resource management, and query optimization.
+
+## Fault-tolerance
+
+There are now two levels of fault-tolerance into Trino:
+
+### Query-level Retry
+
+By maintaining a deduplication buffer with the ability to spill to external
+storage on the coordinator, Trino supports auto restart of a full query in case
+of transient errors. This enables auto failure recovery for interactive and
+short-running batch workloads, without the need for users to intervene.
+
+### Task-level Retry
+
+Task-level retry is supported via storing intermediate data produced by tasks to
+external storage, such as replicated RAM/SSD/HDD for persistence.
+This is also called exchange spooling. This allows restart of a single task in
+case of failure, assuming deterministic partitioning at exchange boundaries, and
+therefore providing granular failure recovery.
+
+This new fault-tolerant execution paradigm presents challenges in
+terms of correctness and performance in the following aspects:
+
+#### Task Scheduling
+
+To ensure correctness, we added task scheduling support for fault-tolerant
+execution to ensure:
+
+1. **Determinism**: In case of failure, a task must be restarted with the same 
+input partition or set of splits.
+2. **Restartability**: Failure of a task should be propagated to the scheduler,
+and scheduler should be able to reschedule the failed task for re-execution.
+3. **Atomicity**: We need to make sure tasks are atomic and don’t produce any
+side effects. A deduplication buffer on Coordinator is used to deduplicate the
+output produced by different instances of the same task. Meanwhile, for insert
+queries, connectors need to make sure non-committed data is not visible.
+
+#### Exchange Spooling
+
+The intermediate storage for spooling data should be able to buffer, store and
+return it efficiently.
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/exchange-spooling.png)
+
+Currently, we have implemented a filesystem-based approach in the 
+`trino-exchange-filesystem` plugin, supporting AWS S3, Azure Blob
+Storage, Google Cloud Storage, as well as S3-compatible systems. The
+implementation takes full advantage of asynchronous APIs to overlap
+computation and I/O, and performs parallel reads and writes from and to exchange
+storage in parts to maximize throughput. We also opportunistically coalesce and
+split partition files depending on the size. 
+
+The exchange APIs defined in the Trino SPI are designed to be extensible, and allow for the community to build a distributed buffering service to further enhance exchange performance.
+
+With fault tolerance built into Trino, long-running batch queries now can enjoy
```

Comment:
This is not quite true. Stable and predictable completion times depend on many other factors than granular retries (concurrent workloads, resource management, performance of dependent systems, etc).

Also, "landing time" is a very vague term (specifically used within FB), but I don't know if it's in common usage in the industry. Is it defined elsewhere?